content/blog/2019-09-25-making-function-calls-work.md

```diff
@@ -0,0 +1,299 @@
+
+Below is a Bril program that demonstrates the use of function parameters.
+
+    # This program also prints out 100 and exits.
+    main {
+      call print_double 50;
+    }
+    
+    print_double x {
+      v0: int = add x x;
+      print v0;
+    }
+
+#### Return types
+Without the ability to return data, our functions are not very useful. To add support for return types, we first update the grammar for function definitions to optionally take `: type` at the end of the header, and update the rule for `ret` to optionally take a variable name to return. We also overload the `call` operation to be both an effect operation and a value operation, since functions that do not return anything will be effect operations and those that do will be value operations.
+
+To pass a return value back to the caller, we use the following approach. When we handle a `ret` operation, if there is a return value, we add a special variable `_ ret` to the environment and map it to the return value. Note that this will never collide with any existing variables as there is space in the name. Then, if the caller is expecting a return value (which it will in instructions of the form `v: type = call func_name`), it can check for the existence of `_ ret` in the callee's environment. We do some basic type-checking here, comparing return types and variable types, to make the Bril programmer's life easier.
+
+Below is an example program that demonstrates this functionality.
+
+    # Once again, we print 100.
+    main {
+      v0: int = call get_hundred;
+      print v0;
+    }
+    
+    get_hundred: int {  # The ": int" is required.
+      v0: int = const 100;
+      ret v0;
+    }
+
+Since we now have the ability to pass arguments to functions and get return values, we can now write some interesting Bril programs. Below is a Bril program that prints the 10th Fibonacci number.
+
+    # Print fib(10)
+    main {
+        v0: int = const 10;
+        fib10: int = call fib v0;
+        print fib10;
+    }
+
+    # Return true if n <= 1, false otherwise.
+    lte_one n: bool {
+        one: int = const 1;
+        lto: bool = le n one;
+        ret lto;
+    }
+
+    # Return fib(10).
+    fib n: int {
+        base: bool = call lte_one n;
+        br base return continue;
+    return:
+        ret n;
+    continue:
+        one: int = const 1;
+        prev: int = sub n one;
+        prev2: int = sub prev one;
+        fib1: int = call fib prev;
+        fib2: int = call fib prev2;
+        ans: int = add fib1 fib2;
+        ret ans;
+    }
+
+**Note**: To call a function on the right-hand side of an assignment, the function must declare a return type. We choose to enforce this to improve readability. The following program will fail with the error message `function func does not return`.
+
+    # Fails!
+    main {
+      v0: int = call func;
+      print v0;
+    }
+
+    func {
+      v0: int = const 100;
+      ret v0;
```

Comment:
It sort of seems like there are _two_ errors in this program: one on the `call` instruction and one on the `ret` instruction. Would you mind terribly much specifying that it's an error to `ret` a value in a function that doesn't have a declared return type? It doesn't matter whether the interpreter actually checks this, but I think it would be nice to make it clear that this is a "bad" program that you shouldn't write.